src/lem/circuit.rs

```diff
@@ -0,0 +1,899 @@
+                    cs,
+                    &slot,
+                    component_idx,
+                    z_ptr.hash,
+                )?);
+
+                component_idx += 1;
+            }
+
+            // Allocate the image by calling the arithmetic function according
+            // to the slot type
+            let preallocated_img =
+                Self::allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+
+            preallocations.push((preallocated_preimg, preallocated_img));
+        }
+
+        // Then we do the same with dummies for the remaining slots
+        for slot_idx in preallocations.len()..num_slots {
+            let slot = Slot {
+                idx: slot_idx,
+                typ: slot_type,
+            };
+            let mut preallocated_preimg = vec![];
+            for component_idx in 0..slot_type.preimg_size() {
+                preallocated_preimg.push(Self::allocate_preimg_component_for_slot(
+                    cs,
+                    &slot,
+                    component_idx,
+                    F::ZERO,
+                )?);
+            }
+            let preallocated_img =
+                Self::allocate_img_for_slot(cs, &slot, preallocated_preimg.clone(), store)?;
+
+            preallocations.push((preallocated_preimg, preallocated_img));
+        }
+        Ok(preallocations)
+    }
+
+    /// Create R1CS constraints for LEM given an evaluation frame. This function
+    /// implements the STEP 3 mentioned above.
+    ///
+    /// Notably, we recursively pass on a `concrete_path: Boolean` variable, which
+    /// encodes whether we are on a *concrete* or *virtual* path. A virtual path
+    /// is one that wasn't taken during interpretation and thus its frame pointers
+    /// weren't bound. A concrete path means that interpretation went down that
+    /// road and the frame data should be complete for the variables on that path.
+    /// For virtual paths we need to create dummy bindings for the meta pointers
+    /// and relax the implications with false premises. The premise is precicely
+    /// `concrete_path`.
+    ///
+    /// Regarding the slot optimizations, STEP 3 uses information gathered during
+    /// STEPs 1 and 2. So we proceed by first allocating preimages and images for
+    /// each slot and then, as we traverse the LEM, we add constraints to make sure
+    /// that the witness satisfies the arithmetic equations for the corresponding
+    /// slots.
+    pub fn synthesize<F: LurkField, CS: ConstraintSystem<F>>(
+        &self,
+        cs: &mut CS,
+        store: &mut Store<F>,
+        slots_count: &SlotsCounter,
+        frame: &Frame<F>,
+    ) -> Result<()> {
+        let alloc_manager = AllocationManager::default();
+        let mut allocated_ptrs: HashMap<AString, AllocatedPtr<F>> = HashMap::default();
+
+        self.allocate_input(cs, store, frame, &mut allocated_ptrs)?;
```

Comment:
I think the notion of 'input' (which here seems to refer to 'public input' at the constraint-system level) is a bit confused altogether here.

Please look at the `synthesize` method in `nova.rs` to see how this should work for Nova (there are no public inputs, but the interface is a bit different when implementing `StepCircuit` — or else the `synthesize` method for `MultiFrame` in `circuit_frame.rs` to see how this should work for Groth16 (or some other notional backend using the `Circuit` trait rather than `StepCircuit` trait).

Here, you're creating inputs for every frame. This would be okay if you assume Bellperson and rc=1. But then the outputs would also need to be made into 'public inputs'.

Reply:
This is unrelated to constraint-system input. It's allocating the pointers for the frame input

Reply:
Okay, I misunderstood. Can you point me to where/how these allocated input pointers are used? I searched for `.input_vars` in `circuit.rs` and don't see any usage apart from when they are assigned in `allocate_input`.

Reply:
They are used when we need to constrain LEMOPs that use them. For example, if the LEM has the input `expr_in` and it's used in a `LEMOP::Hash2`, then we can recover what we've previously inserted in `allocated_ptrs`

Reply:
Which LEMOPs use them? I tried a naive textual search and couldn't figure it out so I'm asking for a more specific pointer.

Reply:
In this example:
```
#[test]
fn test_simple_all_paths_delta() {
    let lem = lem!(expr_in env_in cont_in {
        let cont_out_terminal: Terminal;
        return (expr_in, env_in, cont_out_terminal);
    })
    .unwrap();

    constrain_test_helper(
        &lem,
        &[Ptr::num(Fr::from_u64(42)), Ptr::char('c')],
        SlotsCounter::default(),
    );
}
```
The `LEMOP::Return` uses `expr_in` and `env_in`. So they need to be in `allocated_ptrs` when we call `g.allocated_ptrs.get`

Reply:
I see… so you are allocating the input on every frame? However, these very inputs will need to have been allocated in the previous frame, right? So it would be better to use the previous.

Reply:
i.e. each frame's output is the next frame's input.

Reply:
Yeah, there's some noise coming from the fact that `synthesize` is receiving a mutable reference to a constraint system. I want to make it produce a new one on every iteration. Then we change that with proper efforts to make it accomodate multi-frame evaluations

Reply:
Ah, scratch the comment above. We'll need to polish the interactions between frames. But that can be a new task

Reply:
But then you're moving far away from any kind of realistic usage. If you're going to implement a method called `synthesize` that synthesizes one frame's worth of circuit, it needs to take a constraint system as an argument. Even if you don't choose to implement either the `Circuit` or `StepCircuit` traits yet, doing otherwise is too much of a contortion to be useful.

Reply:
[Sorry, GH refresh lag means we crossed comments here. Nevertheless, please take note of mine for when you do undertake it.]

Reply:
So, right now we're assuming `rc = 1`, so we cannot link the previous frame allocated output to the input of the next. In another PR we shall deal with this